### PR TITLE
updated buildkite pipeline generation

### PR DESCRIPTION
**Pull request by metaflow**
- fixed build for linux (clang was missing)

- removed /monolithic-.. from build directory - it does not add anything and makes path longer for windows which is not great;

- added env-based configuration to control cache and agent targeting;

- print (s)ccache stats to file not to pullute normal log.

### CODE REVIEW

File: .ci/monolithic-linux.sh [hunk 1]
```diff
@@ -17,14 +17,19 @@ set -ex
...
 MONOREPO_ROOT="${MONOREPO_ROOT:="$(git rev-parse --show-toplevel)"}"
-BUILD_DIR="${BUILD_DIR:=${MONOREPO_ROOT}/build/monolithic-linux}"
-
+BUILD_DIR="${BUILD_DIR:=${MONOREPO_ROOT}/build}"
 rm -rf ${BUILD_DIR}
 
 ccache --zero-stats
-ccache --show-config
+
+if [[ -n "${CLEAR_CACHE:-}" ]]; then
```
Comment (aeubanks) [inline .ci/monolithic-linux.sh:25]: how can you set this?

File: .ci/generate-buildkite-pipeline-scheduled [hunk 2]
```diff
@@ -28,42 +40,42 @@ steps:
...
     timeout_in_minutes: 150
     env:
       CC: 'cl'
       CXX: 'cl'
       LD: 'link'
     commands:
-      - 'C:\\BuildTools\\Common7\\Tools\\VsDevCmd.bat -arch=amd64 -host_arch=amd64'
-      - 'bash .ci/monolithic-windows.sh "clang-tools-extra;flang;libclc;lld;llvm;mlir;polly;pstl" "check-all"'
+    - C:\\BuildTools\\Common7\\Tools\\VsDevCmd.bat -arch=amd64 -host_arch=amd64
+    - bash .ci/monolithic-windows.sh "clang-tools-extra;flang;libclc;lld;llvm;mlir;polly;pstl;clang" "check-all"
```
Comment (aeubanks) [inline .ci/generate-buildkite-pipeline-scheduled:81]: make alphabetical just to be consistent and look nice

File: .ci/generate-buildkite-pipeline-scheduled [hunk 3]
```diff
@@ -28,42 +40,42 @@ steps:
...
 
-  - label: ':linux: x64 Debian'
+  - label: ':linux: Linux x64'
     artifact_paths:
-      - '*_result.json'
-      - 'build/monolithic-linux/test-results.xml'
-    agents:
-      queue: 'linux'
+    - artifacts/**/*
+    - '*_result.json'
```
Comment (aeubanks) [inline .ci/generate-buildkite-pipeline-scheduled:46]: quoting is inconsistent, I believe all of these don't need quotes
(other files as well)

**Review by aeubanks — COMMENTED** [review]
the commit title doesn't make sense, can you make it something like "Various fixes to buildkite pipeline generation"

File: .ci/monolithic-linux.sh (hunk 1 above)
Comment (metaflow) [reply .ci/monolithic-linux.sh:25]: you can do that when creating a build in buildkite UI or triggering through API with "environment" varaibles. It's not available for "normal" users

File: .ci/generate-buildkite-pipeline-scheduled (hunk 3 above)
Comment (metaflow) [reply .ci/generate-buildkite-pipeline-scheduled:46]: Done.

File: .ci/generate-buildkite-pipeline-scheduled (hunk 2 above)
Comment (metaflow) [reply .ci/generate-buildkite-pipeline-scheduled:81]: Done.

**Comment by metaflow** [comment]
Test build https://buildkite.com/llvm-project/experimental-premerge-checks/builds/60

**Comment by joker-eph** [comment]
This seems broken: https://buildkite.com/llvm-project/github-pull-requests/builds/1350#018a8848-f70e-47f4-8aec-0bb48c49a7ca

```
$ trap 'kill -- $' INT TERM QUIT; .ci/generate-buildkite-pipeline-premerge \| buildkite-agent pipeline upload
  | 2023-09-12 07:26:50 INFO   Reading pipeline config from STDIN
  | 2023-09-12 07:26:50 INFO   Updating BUILDKITE_COMMIT to "6de56a7dc4ff114eb36f42f1dbe8c63e557f061b"
  | Files modified:
  | mlir/include/mlir/Dialect/Affine/IR/AffineOps.h
  | mlir/include/mlir/Dialect/Affine/IR/AffineOps.td
  | mlir/include/mlir/IR/SymbolInterfaces.td
  | mlir/lib/Dialect/Affine/IR/AffineOps.cpp
  | mlir/test/Dialect/Affine/invalid.mlir
  | mlir/test/Dialect/Affine/load-store-invalid.mlir
  | Directories modified:
  | mlir
  | 2023-09-12 07:26:50 FATAL  Pipeline parsing of "(stdin)" failed: line 5: did not find expected alphabetic or numeric character
  | 🚨 Error: The command exited with status 1
```


This seems to match the ` - '*_result.json'`  replaced with ` - *_result.json`

**Comment by metaflow** [comment]
thanks for the heads up @joker-eph ! I should be fixed now but you need to rebase your PR to see the change. I guess noop commit update should do

**Comment by metaflow** [comment]
ah you've merged it already :smile:

**Comment by ldionne** [comment]
@metaflow The reason for not adding clang to the monolithic pipeline is that clang now has its own pipeline like libc++. After your patch we will be building and testing clang twice for no reason. Can you remove it again?

**Comment by metaflow** [comment]
@ldionne  w/o clang flang will not compile. Do you also check flang and all dependencies of clang in clang pipeline?

**Comment by ldionne** [comment]
I think it's fine to add `clang` to the set of projects in `LLVM_ENABLE_PROJECTS` for dependency purposes, but we could avoid calling `check-clang`. I guess it's moot cause we're using `check-all` right now.

We don't check `flang` in the Clang pipeline because the Clang pipeline's primary responsibility should be to test Clang, not other downstream projects.

**Comment by joker-eph** [comment]
> I think it's fine to add clang to the set of projects in LLVM_ENABLE_PROJECTS for dependency purposes, but we could avoid calling check-clang. I guess it's moot cause we're using check-all right now.


FYI, the script for premerge generates invocations like:

`'./.ci/monolithic-linux.sh "clang;flang;llvm;mlir" "check-flang check-mlir"'`

So even though `clang` is enabled, it shouldn't be tested I think? (we don't want to test LLVM either when it isn't touched)

**Comment by ldionne** [comment]
The scheduled build script does call `check-all`, but you're right about the script that generates pre-commit CI pipelines.

File: .ci/monolithic-windows.sh [hunk 4]
```diff
@@ -45,4 +51,4 @@ cmake -S ${MONOREPO_ROOT}/llvm -B ${BUILD_DIR} \
       -D CMAKE_CXX_COMPILER_LAUNCHER=sccache
 
 echo "--- ninja"
-ninja -C ${BUILD_DIR} ${targets}
+ninja -C "${BUILD_DIR}" "${targets}"
```
Comment (mtvec) [inline .ci/monolithic-windows.sh:54]: Could it be that adding quotes around `targets` breaks Windows build? For example [here](https://buildkite.com/llvm-project/github-pull-requests/builds/1606#018a8da6-0dda-4d24-a1cf-971e3450e4f8):

```
+ ninja -C C:/ws/src/build 'check-clang check-clang-tools check-flang check-lld check-llvm check-mlir check-polly'
ninja: error: unknown target 'check-clang check-clang-tools check-flang check-lld check-llvm check-mlir check-polly'
```

Comment (metaflow) [reply .ci/monolithic-windows.sh:54]: on it

Comment (metaflow) [reply .ci/monolithic-windows.sh:54]: fixed in #66192 , thanks for the headsup

Comment (mtvec) [reply .ci/monolithic-windows.sh:54]: Thanks for the quick fix!

**Pull request merged**